netlib/tcp.py

```diff
@@ -728,7 +731,18 @@ def convert_to_ssl(self, sni=None, alpn_protos=None, **sslctx_kwargs):
 
     def connect(self):
         try:
-            connection = socket.socket(self.address.family, socket.SOCK_STREAM)
+            if not self.connection:
+                connection = socket.socket(self.address.family, socket.SOCK_STREAM)
+            else:
+                connection = self.connection
+
+            if self.spoof_source_address:
+                try:
+                    if not connection.getsockopt(socket.SOL_IP, 19):
+                        connection.setsockopt(socket.SOL_IP, 19, 1)
+                except socket.error as e:
+                    raise exceptions.ProtocolException(
+                        "Failed to spoof the source address: " + e.strerror)
```

Comment:
Sorry, I think this was a misunderstanding on Slack. I meant a mitmproxy.exceptions.ProtocolException here, which is automatically caught by mitmproxy and printed as an error. We are still in netlib here, so raising a TcpException below is the right thing to do.